api/alerts.go

```diff
@@ -0,0 +1,207 @@
+	End         string `yaml:"end"         json:"end"`
+	IsLive      bool   `yaml:"isLive"      json:"isLive"`
+}
+
+type Alert struct {
+	ID                 string     `yaml:"-"                     json:"id"`
+	Name               string     `yaml:"name"                  json:"name"`
+	Query              HumioQuery `yaml:"query"                 json:"query"`
+	Description        string     `yaml:"description,omitempty" json:"description"`
+	ThrottleTimeMillis int        `yaml:"throttleTimeMillis"    json:"throttleTimeMillis"`
+	Silenced           bool       `yaml:"silenced"              json:"silenced"`
+	Notifiers          []string   `yaml:"notifiers"             json:"notifiers"`
+	LinkURL            string     `yaml:"linkURL"               json:"linkURL"`
+	Labels             []string   `yaml:"labels,omitempty"      json:"labels,omitempty"`
+}
+
+type Alerts struct {
+	client *Client
+}
+
+func (c *Client) Alerts() *Alerts { return &Alerts{client: c} }
+
+func (a *Alerts) List(view string) ([]Alert, error) {
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts", view)
+
+	res, err := a.client.HttpGET(url)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	alerts := []Alert{}
+	jsonErr := json.Unmarshal(body, &alerts)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+
+	return alerts, nil
+}
+
+func (a *Alerts) Update(viewName string, alert *Alert) (*Alert, error) {
+	existingID, err := a.convertAlertNameToID(viewName, alert.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert alert name to id: %v", err)
+	}
+
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts/%s", viewName, existingID)
+
+	jsonStr, err := json.Marshal(alert)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert alert to json string: %v", err)
+	}
+	// Humio requires notifiers to be specified even if no notifier is desired
+	if alert.Notifiers == nil {
+		alert.Notifiers = []string{}
+	}
+	res, postErr := a.client.HttpPUT(url, bytes.NewBuffer(jsonStr))
+	if postErr != nil {
+		return nil, fmt.Errorf("could not add alert in view %s with name %s, got: %v", viewName, alert.Name, postErr)
+	}
+
+	body, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		log.Fatal(readErr)
+	}
+
+	resAlert := Alert{}
+	jsonErr := json.Unmarshal(body, &resAlert)
+	if jsonErr != nil {
+		log.Fatal(jsonErr)
+	}
+	return &resAlert, nil
+}
+
+func (a *Alerts) Add(viewName string, alert *Alert, updateExisting bool) (*Alert, error) {
+	nameAlreadyInUse, err := a.alertNameInUse(viewName, alert.Name)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine if alert name is in use: %v", err)
+	}
+	if nameAlreadyInUse {
+		if updateExisting == false {
+			return nil, fmt.Errorf("alert with name %s already exists", alert.Name)
+		}
+		return a.Update(viewName, alert)
+	}
+
+	url := fmt.Sprintf("api/v1/repositories/%s/alerts/", viewName)
```

Comment:
There is some duplication between this `Add()` and the `Update()`. Perhaps there should be a private method that handles the code from this line down which is then called by these functions.